File: frontend/lib/minio.ts
import { Client } from 'minio';
import mime from 'mime-types';
import { Readable } from 'stream';
import { v4 as uuidv4 } from 'uuid';

const MINIO_ENDPOINT = process.env.MINIO_ENDPOINT || 'minio:9000';
const MINIO_ACCESS_KEY = process.env.MINIO_ACCESS_KEY || 'minioadmin';
const MINIO_SECRET_KEY = process.env.MINIO_SECRET_KEY || 'minioadmin';

// Initialize MinIO client
export const getMinioClient = () => {
  return new Client({
    endPoint: process.env.MINIO_ENDPOINT?.split(':')[0] || 'minio',
    port: parseInt(MINIO_ENDPOINT.split(':')[1]) || 9000,
    useSSL: false,
    accessKey: MINIO_ACCESS_KEY,
    secretKey: MINIO_SECRET_KEY,
  });
};

// Set bucket policy
const setBucketPolicy = async (minioClient, bucketName) => {
  const policy = {
    Version: '2012-10-17',
    Statement: [
      {
        Effect: 'Allow',
        Principal: { AWS: ['*'] },
        Action: ['s3:GetObject'],
        Resource: [`arn:aws:s3:::${bucketName}/*`],
      },
    ],
  };

  const policyJson = JSON.stringify(policy);
  await minioClient.setBucketPolicy(bucketName, policyJson);
};

// Create bucket if it doesn't exist
export const createBucket = async (minioClient, bucketName) => {
  const exists = await minioClient.bucketExists(bucketName);
  if (!exists) {
    await minioClient.makeBucket(bucketName);
    await setBucketPolicy(minioClient, bucketName);
  }
};

export const uploadMedia = async (fileContent, filename:string|null, bucketName = 'media') => {
  try {
    if (!fileContent) {
      throw new Error('File content is required.');
    }

    const minioClient = getMinioClient();

    // Ensure the bucket exists
    await createBucket(minioClient, bucketName);

    // Generate a filename if none is provided
    if (!filename) {
      filename = `${uuidv4()}.jpg`; // Default to .jpg
    }

    // Detect content type
    const contentType = mime.lookup(filename) || 'application/octet-stream';

    // Convert fileContent to a readable stream
    const fileStream = new Readable();
    fileStream._read = () => {};
    fileStream.push(fileContent);
    fileStream.push(null);

    // Upload file to MinIO
    await minioClient.putObject(bucketName, filename, fileStream, fileContent.length, {
      'Content-Type': contentType,
    });

    // Generate and return the public URL
    return `http://${MINIO_ENDPOINT}/${bucketName}/${filename}`;
  } catch (error) {
    console.error('Error uploading to MinIO:', error.message || error);
    return null;
  }
};

File: backend (old)/app/minio.py
from minio import Minio
import os
import json

def get_minio_client():
    minio_endpoint = os.getenv('MINIO_ENDPOINT', 'minio:9000')
    minio_access_key = os.getenv('MINIO_ACCESS_KEY')
    minio_secret_key = os.getenv('MINIO_SECRET_KEY')
    return Minio(minio_endpoint, access_key=minio_access_key, secret_key=minio_secret_key, secure=False)

def set_bucket_policy(minio_client, bucket_name):
    policy = {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Principal": {"AWS": ["*"]},
                "Action": ["s3:GetObject"],
                "Resource": [f"arn:aws:s3:::{bucket_name}/*"]
            }
        ]
    }
    policy_json = json.dumps(policy)
    minio_client.set_bucket_policy(bucket_name, policy_json)


def create_bucket(client, bucket_name):
    if not client.bucket_exists(bucket_name):
        client.make_bucket(bucket_name)
        set_bucket_policy(client, bucket_name)